### PR TITLE
Enable modification of transaction object before it gets passed to reporters

### PR DESCRIPTION
**Pull request by honzajavorek**
#### :rocket: Why this change?

Users should be able to modify `transaction.test` contents before it's sent to reporters. Inability to modify the test violates Dredd's life cycle as it was designed and documented.

As a side effect, ensuring this works enables users to sanitize sensitive data in hooks before they get sent into reporters and thus eventually disclosed.

#### :memo: Related issues and Pull Requests

- Closes https://github.com/apiaryio/dredd/issues/623
- Closes https://github.com/apiaryio/dredd/issues/421
- Closes https://github.com/apiaryio/dredd/issues/544
- Related to https://github.com/apiaryio/dredd/pull/640

#### :white_check_mark: What didn't I forget?

- [x] To write tests
- [x] To put [Conventional Changelog](https://dredd.readthedocs.io/en/latest/contributing/#sem-rel) prefixes in front of all my commits
- [x] To add docs

### CODE REVIEW

File: src/transaction-runner.coffee [hunk 1]
```diff
@@ -376,35 +376,36 @@ class TransactionRunner
...
+
+        transaction.results ?= {}
+        transaction.results.general ?= []
+        transaction.results.general.results ?= []
+        transaction.results.general.results.push({severity: 'error', message})
+
+        transaction.test.status = 'fail'
+        transaction.test.message = message
+        transaction.test.results ?= {}
+        for key, value of transaction.results when key isnt 'version'
```
Comment (miiila) [inline src/transaction-runner.coffee]: Maybe some comment what's going on would be useful...

File: src/transaction-runner.coffee [hunk 2]
```diff
@@ -157,9 +157,9 @@ class TransactionRunner
               if key != 'version'
                 data['test']['results'][key] ?= {}
                 data['test']['results'][key]['results'] ?= []
-                data['test']['results'][key]['results'] =
-                  data['test']['results'][key]['results'].concat value
+                data['test']['results'][key]['results'] = data['test']['results'][key]['results'].concat value
```
Comment (miiila) [inline src/transaction-runner.coffee]: NIT: concat(value)

File: src/transaction-runner.coffee [hunk 3]
```diff
@@ -183,6 +182,7 @@ class TransactionRunner
       origin: transaction.origin
       startedAt: transaction.startedAt # number in ms (UNIX timestamp * 1000 precision)
       request: transaction.request
+    logger.debug('Transaction runner emits \'test error\' because of an error in hooks, runHooksForData()', transaction.name, error)
```
Comment (miiila) [inline src/transaction-runner.coffee]: Same as above, I don't understand what is `functionName()` after comma. Applies to all similar outputs in this commit.

File: src/transaction-runner.coffee [hunk 4]
```diff
@@ -157,9 +157,9 @@ class TransactionRunner
               if key != 'version'
                 data['test']['results'][key] ?= {}
                 data['test']['results'][key]['results'] ?= []
-                data['test']['results'][key]['results'] =
-                  data['test']['results'][key]['results'].concat value
+                data['test']['results'][key]['results'] = data['test']['results'][key]['results'].concat value
 
+            logger.debug('Transaction runner emits \'test fail\' for test failed by custom assertions in hooks, emitResult()')
```
Comment (miiila) [inline src/transaction-runner.coffee]: I don't understand part `in hooks, emitResult()`.

File: src/transaction-runner.coffee [hunk 5]
```diff
@@ -375,38 +355,38 @@ class TransactionRunner
...
+  imposeFailure: (transaction, reason) ->
+    transaction.results ?= {}
+    transaction.results.general ?= []
+    transaction.results.general.results ?= []
+    transaction.results.general.results.push({severity: 'error', message: reason})
+
+    transaction.test.status = 'fail'
+    transaction.test.message = reason
+    transaction.test.results ?= {}
+    for key, value of transaction.results when key isnt 'version'
```
Comment (miiila) [inline src/transaction-runner.coffee]: I can see this for multiple times, would think of some helper.

File: test/unit/transaction-runner-test.coffee [hunk 6]
```diff
@@ -450,16 +450,16 @@ describe 'TransactionRunner', ->
...
 
       afterEach ->
-        configuration.emitter.emit.restore()
         configuration.options['method'] = []
+        runner.skipTransaction.restore()
 
       it 'should only perform those requests', (done) ->
         runner.executeTransaction transaction, ->
-          assert.ok configuration.emitter.emit.calledWith 'test skip'
+          assert.ok runner.skipTransaction.called
```
Comment (miiila) [inline test/unit/transaction-runner-test.coffee]: NIT: assert.ok(runner.skipTransaction.called)

**Comment by honzajavorek** [comment]
Thanks! Addressed some in 6c4e44d.

**Comment by w-vi** [comment]
Just a question how the error-ing  hooks work in the language hooks like ruby, python etc?

**Comment by honzajavorek** [comment]
@w-vi If hooks crash, Dredd emits `test error` with `transaction.test` object. There's no way to prevent it, so you have to use the power of your hook language and catch all errors by `try/catch` or similar constructs and turn "errorring" hook into "manually failing" hook. There are examples for that use case and two tests - one proves that erroring hook leaks sensitive data and one proves that user can handle this situation gracefully.

**Comment by netmilk** [comment]
Good job! Accepted and merging.

**Pull request merged**